### DISCUSSION TITLE
Zero Downtime Recommendation?

### DISCUSSION THREAD

**Discussion by dm17** [containrrr/watchtower#600] [Q&A] [answered]
I've already been doing manually what watchtower automates because it is quite easy in my configuration. However, do you have recommendations for when we want traffic routed to the old container(s) until the new version of each container is up? This would allow for zero downtime when running infrastructure with docker-compose... Would just need to take some care with docker-compose container dependencies due to use of databases, I think.

**Comment by victorcmoura** [comment, marked as answer]
Hi, @dm17

Personally, when I need zero downtime, I go for a different orchestration strategy other than docker-compose. As high availability involves more than keeping things running during an update, a more robust orchestrator might help you with this situation.

**Reply by dm17** [reply]
Too much added complexity. Not going to run swarm when I'm 99% the way there.

**Reply by simonaronsson** [reply]
We do not have any suggested strategy for zero downtime updates, nor is it anything I think watchtower would be capable of handling currently. As @victorcmoura so excellently points out, I'd likely go for a different orchestration strategy in that case, like Kubernetes och Swarm. 

If the community has any suggestion on how to approach this, we'd be happy to hear about it! One thing I can think of is to have multiple instances defined in your compose file, and use different watchtower instances, with differing intervals, to monitor them and use a load balancer in front of them to shape the traffic according to your needs.

**Comment by dm17** [comment]
Thanks for opening it up to suggestions... Just seeing a bit of brainstorming would make me happy. K8 is too heavy/complex for me - just not interested.

What about the logic of just not switching the networking over to the new (version of the) container until it is pingable - and then stopping the old instance?

**Reply by victorcmoura** [reply]
That's exactly what more robust orchestrators do. To implement this solution, we would need:
- A health check mechanism, that could be used currently to at least indicate whether an update was healthy or not from an application perspective. So I see it as a great feature for now.
- A persistency mechanism to keep track of which image is a healthy image and, therefore, safe to deploy. This is my biggest concern, as our goal has been to make a lightweight, standalone solution to update docker containers. Requiring a sidecar database, for instance, would kinda mess with this philosophy.

I have been thinking of a solution that doesn't require extra persistence and I would like to share with you so maybe we can see if it is feasible. Watchtower would have to be able to do the following pipeline:
1. Download the new image and tag it with a different name (I'll call it `image:watchtower-tmp-image`)
2. Run the container with `image:watchtower-tmp-image`
3. Do the health checks and abort if something wrong is identified in this stage
4. Given that the health check returned a good state, retag the image with its original tag (`image:latest`, for example)
5. Remap networking to include the newest container as the main one
6. Stop the older container

In theory, it sounds magical. In practice, we would have to deal with the following:
- How to pull an image with a different tag through [Moby](https://github.com/moby/moby/) (actual name of Docker Engine: https://github.com/docker/docker)?
  - Is it needed to retag the image? Can we work with an image id, or some other identifier? How can we guarantee that a broken image won't substitute a previous, validated, healthy image during the pull process?
- How to run doubled containers, with possible bindings to the same host port, possible shared volumes, and an infinity of other dependencies without influencing how the older container is running?
- How to remap docker networking through [Moby](https://github.com/moby/moby/)?

The only certainty is that it would be a huge project by itself haha

**Reply by dm17** [reply]
I guess that could be reduced a bit assuming we just use local images / already tagged images. I'm not even saying this should be a feature of watchtower, but I think this is a good place to discuss because if I brought this up in with docker-compose devs, then I think it would be closed and/or locked.

Swarm & k8 are too complex (for myself & many others), and "docker-compose is not for production" seems to lack evidence. If there were a way to use docker-compose without downtime, then I think many more would. Of course many will say, "well that doesn't mean it is high availability!" I would break that down, however. If we launch docker-compose with a env var using a git SHA (on which my images are built), then I can easily roll back to the previous set of images. So if the new images don't fail a test, then they'll be launched successfully without downtime. If the images do fail a test, then we don't upgrade to them. If the images pass a test and still fail in production, then we can just roll back to the previous images with very little downtime (which will happen even in the most advanced HA setups that lack the test to detect the issue).

Does that make sense to everyone else?

**Reply by victorcmoura** [reply]
> "docker-compose is not for production" seems to lack evidence

Keep in mind that your sense of production may differ from other engineers'. In many cases, production systems require a level of scalability and fault tolerance that docker-compose's toolset simply isn't enough. Dealing with zero downtime (specifically about zero downtime when updating images) is something that even [docker-compose's official docs on production deployment](https://docs.docker.com/compose/production/) doesn't guarantee by default. 

Blue-green deployment (as you suggested) is a solution to a problem faced by many engineers. Plenty of examples are available out there and could be a great material to help you push this feature in your workspace. Here is an example: https://coderbook.com/@marcus/how-to-do-zero-downtime-deployments-of-docker-containers/. But notice that something that works isn't always the best maintainable solution, this type of decision is what makes us engineers still needed out there.

Now, speaking as a collaborator of Watchtower, to run multiple instances (one for each deployment env: blue and green) is not currently supported. I'll soon do the required changes in https://github.com/containrrr/watchtower/pull/511 and it should be made available in future releases.

**Reply by dm17** [reply]
Thanks! I will definitely be trying it out with those features. For now, a few seconds of downtime is reasonable for my purposes.
